### ISSUE TITLE
Allow for a CDI style @Produces feature

### ISSUE THREAD

**Issue by glassfishrobot** [eclipse-ee4j/glassfish-hk2#240]
While Factories are ok for having user created stuff, @Produces is more convenient in many cases simply because you generally do not need to create a whole new service just to have code the produces some service or another.

This question on stack-exchange also asked for such a feature:

[http://stackoverflow.com/questions/22856997/hk2-equivalent-of-provides-in-guice-for-jersey-2](http://stackoverflow.com/questions/22856997/hk2-equivalent-of-provides-in-guice-for-jersey-2)

**Comment by michaelhixson** [comment]
If I were to submit a PR for this, would any of this project's maintainers review it and possibly merge it?

I implemented support for `@Provides` in one of my projects and I'm really liking it so far.  I didn't have to modify hk2 -- I was able to package this feature into a `DynamicConfigurationListener` -- so I could theoretically see it being in `hk2-extras` or in its own experimental package.
